### PR TITLE
[*] CORE: Refactor Category::getProducts methods

### PR DESCRIPTION
**Pull request by gskema**
I wanted to do some more optimizations, but I found out that someone has already refractored that part :)

### CODE REVIEW

File: classes/Category.php [hunk 1]
```diff
@@ -645,9 +651,6 @@ public function getProducts($id_lang, $p, $n, $order_by = null, $order_way = nul
 		if ($order_by == 'price')
 			$order_by = 'orderprice';
 
-		if (!Validate::isBool($active) || !Validate::isOrderBy($order_by) || !Validate::isOrderWay($order_way))
```
Comment (jeromenadaud) [inline classes/Category.php]: Hi, Why removing Validate::isBool($active) ?

Comment (gskema) [inline classes/Category.php]: - I think it's redundant (and also too strict): `$active` var is only used in ternary statement in lines 662, 696 and not inserted into SQL itself. The value is only evaluated as bool, we then can delete the unnecessary `die()` statement.

Comment (jeromenadaud) [inline classes/Category.php]: So maybe just cast active in ternary ?

Comment (gskema) [inline classes/Category.php]: Personally, I don't think it's necessary, but what do you prefer?

**Comment by jeromenadaud** [comment]
it's ok for me, can you pleas update your fork ?
Thank you

File: classes/Category.php [hunk 2]
```diff
@@ -637,39 +637,56 @@ public function getSubCategories($id_lang, $active = true)
...
+					'.Shop::addSqlAssociation('product', 'p').'
+					LEFT JOIN `'._DB_PREFIX_.'category_product` cp ON p.`id_product` = cp.`id_product`
+					WHERE cp.`id_category` = '.(int)$this->id.
+				($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '').
+				($active ? ' AND product_shop.`active` = 1' : '').
+				($id_supplier ? 'AND p.id_supplier = '.(int)$id_supplier : '');
 
-		if ($p < 1) $p = 1;
+			return (int)Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue($sql);
+		}
```
Comment (gskema) [inline classes/Category.php]: - Moved this to front, because it's doesn't need to generate `order_by`

File: classes/Category.php [hunk 3]
```diff
@@ -637,39 +637,56 @@ public function getSubCategories($id_lang, $active = true)
...
 		if (!$context)
 			$context = Context::getContext();
+
 		if ($check_access && !$this->checkAccess($context->customer->id))
 			return false;
 
-		$front = true;
-		if (!in_array($context->controller->controller_type, array('front', 'modulefront')))
-			$front = false;
+		$front = in_array($context->controller->controller_type, array('front', 'modulefront'));
```
Comment (gskema) [inline classes/Category.php]: - `in_array` returns bool, no need to assign it ourselves.

File: classes/Category.php [hunk 4]
```diff
@@ -687,32 +704,16 @@ public function getProducts($id_lang, $p, $n, $order_by = null, $order_way = nul
...
-					FROM `'._DB_PREFIX_.'product` p
-					'.Shop::addSqlAssociation('product', 'p').'
-					LEFT JOIN `'._DB_PREFIX_.'category_product` cp ON p.`id_product` = cp.`id_product`
-					WHERE cp.`id_category` = '.(int)$this->id.
-					($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '').
-					($active ? ' AND product_shop.`active` = 1' : '').
-					($id_supplier ? 'AND p.id_supplier = '.(int)$id_supplier : '');
-			return (int)Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue($sql);
-		}
+		$nb_days_new_product = Configuration::get('PS_NB_DAYS_NEW_PRODUCT');
```
Comment (gskema) [inline classes/Category.php]: - Single call to `Configuration::get` + shorter line

File: classes/Category.php [hunk 5]
```diff
@@ -687,32 +704,16 @@ public function getProducts($id_lang, $p, $n, $order_by = null, $order_way = nul
...
-		$sql = 'SELECT p.*, product_shop.*, stock.out_of_stock, IFNULL(stock.quantity, 0) as quantity'.(Combination::isFeatureActive() ? ', IFNULL(product_attribute_shop.id_product_attribute, 0) id_product_attribute,
+		$sql = 'SELECT p.*, product_shop.*, stock.out_of_stock, IFNULL(stock.quantity, 0) AS quantity'.(Combination::isFeatureActive() ? ', IFNULL(product_attribute_shop.id_product_attribute, 0) AS id_product_attribute,
 					product_attribute_shop.minimal_quantity AS product_attribute_minimal_quantity' : '').', pl.`description`, pl.`description_short`, pl.`available_now`,
 					pl.`available_later`, pl.`link_rewrite`, pl.`meta_description`, pl.`meta_keywords`, pl.`meta_title`, pl.`name`, image_shop.`id_image` id_image,
 					il.`legend` as legend, m.`name` AS manufacturer_name, cl.`name` AS category_default,
 					DATEDIFF(product_shop.`date_add`, DATE_SUB("'.date('Y-m-d').' 00:00:00",
-					INTERVAL '.(Validate::isUnsignedInt(Configuration::get('PS_NB_DAYS_NEW_PRODUCT')) ? Configuration::get('PS_NB_DAYS_NEW_PRODUCT') : 20).'
-						DAY)) > 0 AS new, product_shop.price AS orderprice
+					INTERVAL '.(int)$nb_days_new_product.' DAY)) > 0 AS new, product_shop.price AS orderprice
```
Comment (gskema) [inline classes/Category.php]: - cast to integer just in case

File: classes/Category.php [hunk 6]
```diff
@@ -747,13 +748,14 @@ public function getProducts($id_lang, $p, $n, $order_by = null, $order_way = nul
 			LIMIT '.(((int)$p - 1) * (int)$n).','.(int)$n;
 
 		$result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql, true, false);
-		if ($order_by == 'orderprice')
-			Tools::orderbyPrice($result, $order_way);
 
 		if (!$result)
 			return array();
```
Comment (gskema) [inline classes/Category.php]: - If empty, skip `Tools::orderbyPrice` call and return immediately.

**Comment by gskema** [comment]
I synced my branch. I noticed some new things that could be improved, so I updated again. Please check. Thanks

**Comment by jeromenadaud** [comment]
Hi,

Thank you for your contribution.

Best regards.

**Pull request merged**